### DISCUSSION TITLE
Random generation of points, given a discretized probability density function

### DISCUSSION THREAD

**Discussion by byba1990** [pyvista/pyvista#2703] [Q&A]
Hi all, 

with respect the probability distribution that I presented [here](https://github.com/pyvista/pyvista/discussions/1986), I would like to ask another suggestion, please. In my work, I need to carry out a random generation of points (particles), considering the probability density distribution (discretized) that I obtained previously. In particular, a single point (particle) corresponds to a specific position (x,y,z) inside a cell. Do you have any suggestion, please?
Thanks in advance for your time.
Best regards,

Christian

![Schermata da 2022-05-26 16-04-53](https://user-images.githubusercontent.com/45231195/170493336-03027642-72f8-4d3a-af39-3866538f8670.png)

**Comment by MatthewFlamm** [comment]
Here is a sketch of how you can do this with a simpler example:

```py
import pyvista
import numpy as np
grid = pyvista.UniformGrid(dims=(10, 10, 10))
# generate concentrated probabilities
grid["prob"] = np.exp(np.linspace(0, 5, grid.n_cells))
grid["prob"] /= np.sum(grid["prob"])
# find random cells according to probabilities
cells = np.random.choice(list(range(grid.n_cells)), size=200, p=grid["prob"])
# find random points inside cell to place points (more complex cell types will be more complex)
grid_center = grid.cell_centers()
jitter = (2 * np.random.random((200, 3)) - 1) * 0.5
points = pyvista.PolyData(grid_center.points[cells]+ jitter)
# make pretty plot
glyphs = points.glyph(geom=pyvista.Sphere(radius=0.1))
plotter = pyvista.Plotter()
plotter.add_mesh(glyphs)
plotter.add_mesh(grid, style='wireframe')
plotter.show()
```

![image](https://user-images.githubusercontent.com/39341281/170511512-e947f232-e083-423e-aef5-97dd91eaaca5.png)

**Reply by MatthewFlamm** [reply]
How to implement for an arbitrary cell shape?  One thought is to first choose random cells as above.  Then for each cell chosen, pick a random point inside the bounding box of the cell, if the point is inside the cell, accept it.  If not, repeat.  Unfortunately, I don't think pyvista exposes interacting with individual cells.  Assessing whether a point is inside a specific cell can be done through the the `CellBoundary` or `EvaluatePosition` methods of `vtkCell`.

Edit:  This method will be really slow for very skewed meshes, and somewhat slow compared to a deterministic approach like the example.

**Reply by byba1990** [reply]
Dear @MatthewFlamm ,

thank you very much for your answer and your suggestions. Unfortunately, I still have a problem to get points from the probability distribution of the cells, considering that the cells of the mesh are more complex than the proposed example and are 2D cells. Then, it doesn't work also using the bounding box for arbitrary cell shapes, in addition to the fact that I would like to perform a fast procedure. To figure out my specific problem, here below the mesh of a sector of my geometry:

![Schermata da 2022-05-27 13-18-08](https://user-images.githubusercontent.com/45231195/170681756-46abbf67-5ff9-4275-8b61-bb5b393686bd.png)

In particular, I aim to generate points on the surface, with these cells that have a 2D hexahedral shape (surface) of a 3D geometry. 
Referring to the example that you proposed, I guess it is possible to solve the problem as you suggested by: 
1. finding a random cells according to the probabilities (did it as you suggested) 
2. finding random points inside the cell. In your case, you did it with respect the specific cell type. With the particular mesh of my problem, I don't find a way to do it, if not to have the possibility to generate random points by having the collection of points belonging to the specific cell. The random generation of points considering only the boundaries allows to obtain points inside a rectangle, that is still an approximation of the cell shape. Also in this case, the other solution is the acceptance/rejection approach that you suggested, that unfortunately require more computational time.

If there are other suggestions to solve this problem, please, let me know.

Thank you very much for your time.
Best regards,

Christian

**Reply by MatthewFlamm** [reply]
Can you transform your cell type into a unit cube, pick a random point, then transform back?

**Reply by byba1990** [reply]
How can I do this, please?

**Reply by MatthewFlamm** [reply]
This is now outside my area of expertise, but I _think_ you can use perspective transformation to transform a wedge shaped hexahedral, which is what I think you have, into a cube or vice versa.  I don't know how to accomplish it in Python code.  Maybe someone else can help here.

**Reply by MatthewFlamm** [reply]
Thinking about this some more, if you use perspective transformation, then you won't achieve uniform density within a cell when transforming back to the original cell shape.

Since your mesh seems to be very regular in a cylindrical coordinate system, you should be able to take take a similar approach that I used for a cubical mesh.  First, find which cell to add a point.  Second, map cell id to it's limits of radius `R`, sector angles `theta`, and thickness position `z`.  This could be precomputed.  Generate random point between `R1` and `R2` while preserving uniform density, and a uniformly random `theta` and `z`.
